[RISCV] Widen i1 AnyOf reductions

lukel97 opened the pull request:
With EVL tail folding an AnyOf reduction will end up emitting an i1 vp.merge.

Unfortunately due to RVV not containing any tail undisturbed mask instructions, an i1 vp.merge will get expanded to a lengthy sequence:

```asm
  vsetvli a1, zero, e64, m1, ta, ma
  vid.v v10                        
  vmsltu.vx v10, v10, a0           
  vmand.mm v9, v9, v10             
  vmandn.mm v8, v8, v9             
  vmand.mm v9, v0, v9              
  vmor.mm v0, v9, v8               
```

This addresses this by matching this specific AnyOf pattern in RISCVCodegenPrepare and widening it from i1 to i8, which will end up producing a single masked i8 vor.vi inside the loop:

```llvm
loop:                                                                      
  %phi = phi <vscale x 4 x i1> [ zeroinitializer, %entry ], [ %rec, %loop ]
  %cmp = icmp ...                                                                                          
  %rec = call <vscale x 4 x i1> @llvm.vp.merge(%cmp, true, %phi, %evl)     
```

```llvm
loop:                                                                      
  %phi = phi <vscale x 4 x i8> [ zeroinitializer, %entry ], [ %rec, %loop ]
  %cmp = icmp ...                             
  %rec = call <vscale x 4 x i8> @llvm.vp.merge(%cmp, true, %phi, %evl)     
  %trunc = trunc <vscale x 4 x i8> %rec to <vscale x 4 x i1>               
```

I ended up adding this in RISCVCodegenPrepare instead of the LoopVectorizer itself since it would have required adding a target hook.

It may also be possible to generalize this to other i1 vp.merges in future. 

Normally the trunc will be sunk outside of the loop. But it also doesn't check to see if all the non-phi users of the vp.merge are outside of the loop: If there are in-loop users this still seems to be profitable, see the test diff in `@widen_anyof_rdx_use_in_loop`

Fixes #132180

lukel97 committed:
Precommit tests

lukel97 committed:
[RISCV] Widen i1 AnyOf reductions

preames commented:
I think you have a missing simplification in your input.  Can't this:
```
%phi = phi <vscale x 4 x i1> [ zeroinitializer, %entry ], [ %rec, %loop ]
  %cmp = icmp ...                                                          
  %or = or <vscale x 4 x i1> %phi, %cmp                                    
  %rec = call <vscale x 4 x i1> @llvm.vp.merge(TrueMask, %or, %phi, %evl)   
```

Be simplified to:
```
%phi = phi <vscale x 4 x i1> [ zeroinitializer, %entry ], [ %rec, %loop ]
  %cmp = icmp ...                                                          
  %rec = call <vscale x 4 x i1> @llvm.vp.merge(%cmp, TrueMask, %phi, %evl)   
```

lukel97 commented:
> I think you have a missing simplification in your input. Can't this:
> 
> ```
> %phi = phi <vscale x 4 x i1> [ zeroinitializer, %entry ], [ %rec, %loop ]
>   %cmp = icmp ...                                                          
>   %or = or <vscale x 4 x i1> %phi, %cmp                                    
>   %rec = call <vscale x 4 x i1> @llvm.vp.merge(TrueMask, %or, %phi, %evl)   
> ```
> 
> Be simplified to:
> 
> ```
> %phi = phi <vscale x 4 x i1> [ zeroinitializer, %entry ], [ %rec, %loop ]
>   %cmp = icmp ...                                                          
>   %rec = call <vscale x 4 x i1> @llvm.vp.merge(%cmp, TrueMask, %phi, %evl)   
> ```

I think so. The above is just what the loop vectorizer currently generates, and it doesn't look like there's any existing combine for this in InstCombine/VectorCombine. 

We could simplify this in VPlan and get the slightly cheaper costing. I'll see if I can do that first and then return to this PR

topperc commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -103,6 +105,82 @@ bool RISCVCodeGenPrepare::visitAnd(BinaryOperator &BO) {
+    return false;
+
+  auto *Phi = dyn_cast<PHINode>(PhiV);
+  auto *Start = dyn_cast<Constant>(Phi->getIncomingValue(0));
```
I think it is possible to have a phi with no operands so I'm not sure you can assume value 0 exists.

lukel97 commented:
Apologies for the delay, I've rebased this on top of #135017 now that it's landed and simplified the pattern.

lukel97 committed:
Reduce num of phi uses needed to 1

Previously when we were checking for we were matching the or it was also a use on the phi

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:
```diff
@@ -103,6 +105,76 @@ bool RISCVCodeGenPrepare::visitAnd(BinaryOperator &BO) {
+// The trunc will normally be sunk outside of the loop, but even if there are
+// users inside the loop it is still profitable.
+bool RISCVCodeGenPrepare::widenVPMerge(IntrinsicInst &II) {
+  if (!II.getType()->getScalarType()->isIntegerTy(1))
```
Should we check `getPreferredTailFoldingStyle`?

wangpc-pp replied:
No, the option can be overridden by CLI. Forget it.

lukel97 replied:
I think this should always be profitable in any case, even if it doesn't come from say the loop vectorizer. Looks like I need to update the comment about it being folded into the vor.vv though

wangpc-pp reviewed (approved):
LGTM.

lukel97 committed:
Update comment

preames reviewed (approved):
LGTM

lukel97 committed:
Merge branch 'main' into riscv/widen-i1-vpmerge
